File: src/test/java/com/mkreidl/ephemeris/solarsystem/VSOP87DTest.java
package com.mkreidl.ephemeris.solarsystem;

import static org.junit.Assert.assertEquals;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;

import com.mkreidl.ephemeris.geometry.Spherical;
import com.mkreidl.ephemeris.geometry.VSOP87File;

@RunWith(Parameterized.class)
public class VSOP87DTest extends VSOP87Test
{
    @Parameters(name = "{0} -- {1}")
    public static Iterable<Object[]> data()
    {
        return data(VSOP87File.Version.D);
    }

    private final Spherical expectedPos = new Spherical();
    private final Spherical actualPos = new Spherical();
    private final Spherical expectedVel = new Spherical();
    private final Spherical actualVel = new Spherical();

    public VSOP87DTest( VSOP87File.Planet planet, String timeStr, DataSet dataSet )
    {
        super(planet, timeStr, dataSet);
        expectedPos.set( dataSet.coordinates[2], dataSet.coordinates[0], dataSet.coordinates[1] );
        expectedVel.set( dataSet.coordinates[5], dataSet.coordinates[3], dataSet.coordinates[4] );
    }

    @Test
    public void testModel()
    {
        final AbstractModelVsop87.LBR model;
        switch ( planet )
        {
            case MER:
                model = AbstractModelVsop87.getVersionD( AbstractModelVsop87.MERCURY );
                break;
            case VEN:
                model = AbstractModelVsop87.getVersionD( AbstractModelVsop87.VENUS );
                break;
            case EAR:
                model = AbstractModelVsop87.getVersionD( AbstractModelVsop87.EARTH );
                break;
            case MAR:
                model = AbstractModelVsop87.getVersionD( AbstractModelVsop87.MARS );
                break;
            case JUP:
                model = AbstractModelVsop87.getVersionD( AbstractModelVsop87.JUPITER );
                break;
            case SAT:
                model = AbstractModelVsop87.getVersionD( AbstractModelVsop87.SATURN );
                break;
            case URA:
                model = AbstractModelVsop87.getVersionD( AbstractModelVsop87.URANUS );
                break;
            case NEP:
                model = AbstractModelVsop87.getVersionD( AbstractModelVsop87.NEPTUNE );
                break;
            default:
                model = null;
        }
        model.compute( time, actualPos, actualVel );
        actualVel.dst *= 86400;  // Reference values from VSOP test files are given in [dist] per DAY
        actualVel.lon *= 86400;  // Reference values from VSOP test files are given in [dist] per DAY
        actualVel.lat *= 86400;  // Reference values from VSOP test files are given in [dist] per DAY
        assertEquals( expectedPos.dst, actualPos.dst, 1e-10 );
        assertEquals( expectedPos.lon, actualPos.lon, 1e-09 );
        assertEquals( expectedPos.lat, actualPos.lat, 1e-10 );
        assertEquals( expectedVel.dst, actualVel.dst, 1e-8 );
        assertEquals( expectedVel.lon, actualVel.lon, 1e-8 );
        assertEquals( expectedVel.lat, actualVel.lat, 1e-8 );
    }
}

File: src/test/java/com/mkreidl/ephemeris/solarsystem/Vsop87DTest.kt
package com.mkreidl.ephemeris.solarsystem

import com.mkreidl.ephemeris.SECONDS_PER_DAY
import com.mkreidl.ephemeris.solarsystem.vsop87.ModelVsop87
import com.mkreidl.ephemeris.solarsystem.vsop87.d.*
import com.mkreidl.ephemeris.time.Instant
import com.mkreidl.ephemeris.util.Vsop87TestUtil
import com.mkreidl.math.Spherical3
import com.mkreidl.math.Vector3
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.junit.runners.Parameterized.Parameters

@RunWith(Parameterized::class)
class Vsop87DTest(private val planet: Vsop87TestUtil.Planet, private val instant: Instant, coordinates: DoubleArray) {

    private val expectedPos = Spherical3(coordinates[2], coordinates[0], coordinates[1])
    private val expectedVel = Vector3(coordinates[5], coordinates[3], coordinates[4])

    @Test
    fun testModel() {
        val model: ModelVsop87.LBR = when (planet) {
            Vsop87TestUtil.Planet.MER -> MercuryVsop87D()
            Vsop87TestUtil.Planet.VEN -> VenusVsop87D()
            Vsop87TestUtil.Planet.EAR -> EarthVsop87D()
            Vsop87TestUtil.Planet.MAR -> MarsVsop87D()
            Vsop87TestUtil.Planet.JUP -> JupiterVsop87D()
            Vsop87TestUtil.Planet.SAT -> SaturnVsop87D()
            Vsop87TestUtil.Planet.URA -> UranusVsop87D()
            Vsop87TestUtil.Planet.NEP -> NeptuneVsop87D()
            else -> throw IllegalArgumentException("Planet not found")
        }
        var (actualPos, actualVel) = model.compute(instant).spherical
        actualPos = actualPos.reduce()
        assertEquals(expectedPos.dst, actualPos.dst, 1e-9)
        assertEquals(expectedPos.lon, actualPos.lon, 1e-9)
        assertEquals(expectedPos.lat, actualPos.lat, 1e-9)
        // Reference values from VSOP test files are given in [dist] per DAY
        assertEquals(expectedVel.x, actualVel.x * SECONDS_PER_DAY, 1e-9)
        assertEquals(expectedVel.y, actualVel.y * SECONDS_PER_DAY, 1e-9)
        assertEquals(expectedVel.z, actualVel.z * SECONDS_PER_DAY, 1e-9)
    }

    companion object {
        @JvmStatic
        @Parameters(name = "{0} -- {1}")
        fun data() = Vsop87TestUtil.data(Vsop87TestUtil.Version.D)
    }
}
